feat(polys): Add Gaussian domains to Poly

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #19579
Follows on from #15396


#### Brief description of what is fixed or changed

Add the Gaussian domains to Poly so that they are used when appropriate e.g.:
```julia
In [1]: Poly(I*x, x)                                                                                                                           
Out[1]: Poly(I*x, x, domain='ZZ_I')
```

#### Other comments

There are still some problems here most notably this line
https://github.com/sympy/sympy/blob/46469d714a095fde59f9c4d83aaa528c68b0be0f/sympy/polys/euclidtools.py#L996
because `is_negative` does not work in the Gaussian domains.

Also there needs to be an `lcm` function for Gaussian domains.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Fix flake8

jksuom commented:
> because is_negative does not work in the Gaussian domains.

This is a headache also in many other domains. For polynomials, it means that the leading coefficient is negative: https://github.com/sympy/sympy/blob/25fbcce5b1a4c7e3956e6062930f4a44ce95a632/sympy/polys/domains/polynomialring.py#L128-L130 It is not really needed in `euclidtools`. For Gaussian domains, I would suggest adding a method always returning False.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add is_negative method to GaussianDomain and some other methods that
turned out to be necessary.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add lcm to GaussianIntegerRing.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add missing methods and correct unify logic for e.g. ZZ_I with QQ.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add GaussianRationalField.denom and fix GaussianInteger.lcm

oscarbenjamin commented:
It seems that roots fails over the Gaussian integers if there is an `I` in the generator:
```julia
In [14]: p = Poly(-2*I*x**15 - 6*I*x**13 - 4*I*x**11 + 4*I*x**9 + 6*I*x**7 + 2*I*x**5, x, domain='ZZ_I')                                       

In [15]: p                                                                                                                                     
Out[15]: Poly(-2*I*x**15 - 6*I*x**13 - 4*I*x**11 + 4*I*x**9 + 6*I*x**7 + 2*I*x**5, x, domain='ZZ_I')

In [16]: roots(p)                                                                                                                              
Out[16]: {-1: 1, 0: 5, 1: 1, -ⅈ: 4, ⅈ: 4}

In [17]: p2 = p.subs(x, exp(I*x/2))                                                                                                            

In [18]: p2                                                                                                                                    
Out[18]: 
Poly(-2*I*(exp(I*x/2))**15 - 6*I*(exp(I*x/2))**13 - 4*I*(exp(I*x/2))**11 + 4*I*(exp(I*x/2))**9 + 6*I*(exp(I*x/2))**7 + 2*I*(exp(I*x/2))**5, e
xp(I*x/2), domain='ZZ_I')

In [19]: roots(p2)                                                                                                                             
Out[19]: {0: 5}
```
This leads to a test failure because:
```julia
In [1]: solve(sin(x) + tan(x))                                                                                                                 
Out[1]: []
```
It seems that solve gets there by rewriting as `tan` before rewriting as `exp` which is suboptimal:
```julia
In [15]: expr = sin(x)+tan(x)                                                                                                                  

In [16]: expr.rewrite(tan).rewrite(exp).as_numer_denom()[0]                                                                                    
Out[16]: 
                                         5⋅ⅈ⋅x                                                                      3⋅ⅈ⋅x
                         2               ─────                  ⎛            2                  2     ⎞             ─────
    ⎛     ⅈ⋅x⎞ ⎛ ⅈ⋅x    ⎞  ⎛ 2⋅ⅈ⋅x    ⎞    2       ⎛     2⋅ⅈ⋅x⎞ ⎜  ⎛     ⅈ⋅x⎞   ⅈ⋅x   ⎛ ⅈ⋅x    ⎞   ⅈ⋅x⎟ ⎛ ⅈ⋅x    ⎞    2  
2⋅ⅈ⋅⎝1 - ℯ   ⎠⋅⎝ℯ    + 1⎠ ⋅⎝ℯ      + 1⎠⋅ℯ      + ⅈ⋅⎝1 - ℯ     ⎠⋅⎝- ⎝1 - ℯ   ⎠ ⋅ℯ    + ⎝ℯ    + 1⎠ ⋅ℯ   ⎠⋅⎝ℯ    + 1⎠⋅ℯ     

In [17]: expr.rewrite(exp).as_numer_denom()[0]                                                                                                 
Out[17]: 
    ⎛     2⋅ⅈ⋅x⎞  2⋅ⅈ⋅x     ⎛ 2⋅ⅈ⋅x    ⎞ ⎛ 2⋅ⅈ⋅x    ⎞  ⅈ⋅x
2⋅ⅈ⋅⎝1 - ℯ     ⎠⋅ℯ      - ⅈ⋅⎝ℯ      - 1⎠⋅⎝ℯ      + 1⎠⋅ℯ  
```

oscarbenjamin commented:
On master we have:
```julia
In [1]: f = x/((x - 2)*(x + I))                                                                                                                

In [2]: f                                                                                                                                      
Out[2]: 
       x       
───────────────
(x - 2)⋅(x + ⅈ)

In [3]: apart(f)                                                                                                                               
Out[3]: 
 1 + 2⋅ⅈ    2⋅(-2 + ⅈ)
───────── - ──────────
5⋅(x + ⅈ)   5⋅(x - 2)
```
This PR gives:
```julia
In [5]: apart(f)                                                                                                                               
Out[5]: 
       ⅈ⋅x       
─────────────────
(x - 2)⋅(ⅈ⋅x - 1)
```

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Update tests

jksuom commented:
> return roots_binomial(f)

This is only run when `f.length() == 2`. In this case it should be 3.

jksuom commented:
It seems that `_try_heuristics` is run twice. This will happen on the second time.
```
> /home/jks/test/sympy/sympy/polys/polyroots.py(942)_try_heuristics()
-> if f.length() == 2:
(Pdb) p f
Poly(x**2 + 1/3 + 2**(1/3)/(3*(2 + 27*I + 3*sqrt(3)*sqrt(-27 + 4*I))**(1/3)) + 2**(2/3)*(2 + 27*I + 3*sqrt(3)*sqrt(-27 + 4*I))**(1/3)/6, x, domain='EX')
```

jksuom commented:
It looks like this is coming from Cardano's formulas.

oscarbenjamin commented:
> `roots` hangs on a Poly over the Gaussian integers:

It doesn't "hang" it's just slow (something like 10 minutes). AFAICT the time is spent in expand in this line:
https://github.com/sympy/sympy/blob/25fbcce5b1a4c7e3956e6062930f4a44ce95a632/sympy/polys/polyroots.py#L376-L377

The expressions being expanded take just as long to expand on master but on master we take a different path somehow meaning that it isn't necessary to expand these expressions. It's possible that removing that call to expand is a good idea but the call in itself is not the cause of the slowdown in this PR because it isn't called in the same way on master.

jksuom commented:
> domain='EX'

This is the most common reason for bad performance. There should be a better way of adjoining algebraic numbers like `sqrt(3)` and `sqrt(-27 + 4*I)` to `QQ_I`. It looks like `unify` should be improved.

oscarbenjamin commented:
On master the example is faster because `roots` doesn't enter the `roots_binomial` function and just returns `{}`. The reason `roots` is slow for this case in the PR is that execution now successfully reaches `roots_binomial` which makes some very costly calls to expand. I've profiled this with cProfile/gprof2dot:
![output](https://user-images.githubusercontent.com/1159732/83870100-8bd76880-a725-11ea-87f7-af40628f98c9.png)

oscarbenjamin commented:
Removing the calls to expand makes it much faster:
```diff
diff --git a/sympy/polys/polyroots.py b/sympy/polys/polyroots.py
index 979b27483b..74392ba375 100644
--- a/sympy/polys/polyroots.py
+++ b/sympy/polys/polyroots.py
@@ -373,9 +373,6 @@ def roots_binomial(f):
     base = -cancel(b/a)
     alpha = root(base, n)
 
-    if alpha.is_number:
-        alpha = alpha.expand(complex=True)
-
     # define some parameters that will allow us to order the roots.
     # If the domain is ZZ this is guaranteed to return roots sorted
     # with reals before non-real roots and non-real sorted according
@@ -412,8 +409,8 @@ def roots_binomial(f):
     # compute the roots
     roots, d = [], 2*I*pi/n
     for k in ks:
-        zeta = exp(k*d).expand(complex=True)
-        roots.append((alpha*zeta).expand(power_base=False))
+        zeta = exp(k*d)
+        roots.append(alpha*zeta)
 
     return roots
```
With that it goes from taking 10 minutes to 0.5 seconds:
```julia
In [1]: p = Poly(x**6 + x**4 + I, x)                                                                                              

In [2]: %time ok=roots(p)                                                                                                         
CPU times: user 572 ms, sys: 3.47 ms, total: 576 ms
Wall time: 575 ms
```

jksuom commented:
There is practically no support in `polyroots` for domains other than `ZZ`, `QQ` and then `EX`. Maybe a Gaussian domain should be changed to `EX` at entry for the time being.

oscarbenjamin commented:
> There is practically no support in `polyroots` for domains other than `ZZ`, `QQ` and then `EX`. Maybe a Gaussian domain should be changed to `EX` at entry for the time being.

It could be. The roots returned here (after removing expand calls) are correct though:
```julia
In [16]: [p.subs(x, r).simplify() for r in roots(p)]                                                                              
Out[16]: [0, 0, 0, 0, 0, 0]
```

oscarbenjamin commented:
In this PR if the generator or domain are specified then `Poly` will construct the Poly correctly:
```julia
In [13]: Poly(x**6 + x**4 + I, x)                                                                                                              
Out[13]: Poly(x**6 + x**4 + I, x, domain='ZZ_I')

In [14]: Poly(x**6 + x**4 + I, domain='ZZ_I')                                                                                                  
Out[14]: Poly(x**6 + x**4 + I, x, domain='ZZ_I')
```
However when guessing both generators and domain it will only guess that `I` is part of the domain rather than a generator when passed `extension=True`:
```julia
In [16]: Poly(x**6 + x**4 + I, extension=True)                                                                                                 
Out[16]: Poly(x**6 + x**4 + I, x, domain='ZZ_I')

In [17]: Poly(x**6 + x**4 + I)                                                                                                                 
Out[17]: Poly(x**6 + x**4 + I, x, I, domain='ZZ')
```
Should that last result be changed to return the same as the others?

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Don't use Gaussian domains in roots()

jksuom commented:
> Should that last result be changed to return the same as the others?

That would seem natural to me but maybe not easy to implement.

oscarbenjamin commented:
> > Should that last result be changed to return the same as the others?
> 
> That would seem natural to me but maybe not easy to implement.

Is there any reason to treat `I` as a generator in any situation?

jksuom commented:
> Is there any reason to treat I as a generator in any situation?

Not that I know of.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Use EX domain instead of QQ_I in roots and fix handling of non symbol
generators in roots.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Use AlgebraicField instead of the Gaussian domains for factoring in
apart. The factorisation algorithms are not provided for Gaussian
domains yet.

oscarbenjamin commented on sympy/polys/partfrac.py:
```diff
@@ -116,6 +116,16 @@ def apart(f, x=None, full=False, **options):
                     pass
             return f.xreplace(dict(reps))
 
+    # Factorization over ZZ_I or QQ_I is not supported so we use the
+    # factorization routines for AlgebraicField instead.
+    dom = P.domain
+    if dom.is_ZZ_I or dom.is_QQ_I:
+        if dom.is_ZZ_I:
+            dom = dom.get_field()
+        dom_alg = dom.as_AlgebraicField()
+        P = P.per(P.rep.convert(dom_alg))
+        Q = Q.per(Q.rep.convert(dom_alg))
```
Factorisation over `ZZ_I` and `QQ_I` does not work so I've prevented them from being used in `apart` here. Instead `QQ<I>` will be used.
```julia
In [16]: p1 = Poly(x**2 + 1, x)                                                                                                                

In [17]: p2 = Poly(x**2 + 1, x, domain=QQ_I.as_AlgebraicField())                                                                               

In [18]: p1.factor_list()                                                                                                                      
Out[18]: (1, [(Poly(x**2 + 1, x, domain='ZZ'), 1)])

In [19]: p2.factor_list()                                                                                                                      
Out[19]: (1, [(Poly(x - I, x, domain='QQ<I>'), 1), (Poly(x + I, x, domain='QQ<I>'), 1)])
```
Would it be better to make factorisation over `QQ_I` and `ZZ_I` work?

I wonder if it's possible to make `dup/dmp` treat `QQ_I` as an algebraic field when factorising. I guess that factorising over `ZZ_I` is harder...

jksuom replied:
> Would it be better to make factorisation over `QQ_I` and `ZZ_I` work?

I think that `QQ_I` should be replaced by `QQ<I>` somewhere, maybe in `dup/dmp_ext_factor`.

oscarbenjamin replied:
Would it be possible to add attributes or methods to `QQ_I` so that `dup/dmp` can treat it transparently as an algebraic field without needing any conversion?

Would it make sense to have `QQ_I.is_AlgebraicField` or some other attribute that is common between `QQ_I` and the `AlgebraicField` domains?

It looks like the main attributes needed for an algebraic field are `ext` and `mod`...

jksuom replied:
It looks like at least new versions of `dup/dmp_sqf_norm` would be needed. `QQ_I` should have `conjugate` and the norm of `f` will be `f*cf` where `cf` is `f` with all coefficients conjugated.

oscarbenjamin replied:
Would it make sense for `ZZ_I` to factorise like `QQ_I` or perhaps like `ZZ[I]`?

On master factor can factorise in some cases by treating `I` as a generator:
```julia
In [14]: e = expand((x - I)*(x + 1))                                                                                                           

In [15]: e                                                                                                                                     
Out[15]: 
 2              
x  + x - ⅈ⋅x - ⅈ

In [16]: factor(e)                                                                                                                             
Out[16]: (x + 1)⋅(x - ⅈ)
```
The PR currently gives:
```julia
In [14]: factor(e)                                                                                                                             
Out[14]: 
 2                
x  + x⋅(1 - ⅈ) - ⅈ
```

jksuom replied:
Factorizing polynomials over `ZZ_I` should be done in two steps, factorizing the primitive part over `QQ_I` and the content in `ZZ_I`.

oscarbenjamin replied:
What would `factor_list` return in that case?

At the moment `factor_list` returns a tuple of the content and then a list of factors of the primitive:
```julia
In [21]: Poly(x**2/2 - 2, x).factor_list()                                                                                                     
Out[21]: (1/2, [(Poly(x - 2, x, domain='QQ'), 1), (Poly(x + 2, x, domain='QQ'), 1)])
```
If the content itself factorises where would that go in this data structure?

Is there an implementation for factorisation of gaussian integers anywhere?

Perhaps if not then for now `factor_list` can return the content unfactorised...

jksuom replied:
> Is there an implementation for factorisation of gaussian integers anywhere?

No, but that would not be hard to implement if desired. First factorize the norm in `ZZ` and then factor those prime factors that split in `ZZ_I`.

oscarbenjamin commented:
Constructing a composite domain like `ZZ_I[x]` fails in `construct_domain` because here
https://github.com/sympy/sympy/blob/25fbcce5b1a4c7e3956e6062930f4a44ce95a632/sympy/polys/constructor.py#L118-L124
 we end up with `I` as a generator from `parallel_dict_from_basic`.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Make construct_domain identify composite domains over the Gaussian
integers or rationals e.g. ZZ_I[x] and QQ_I[x] and prevent Poly from
treating I as a generator:

  >>> Poly(x**2 + I)
  Poly(x**2 + I, x, domain='ZZ_I')

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Remove unneeded import

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add properties is_Numerical and is_Exact to ZZ_I and QQ_I. Make it
possible to specify composite domains like ZZ_I[x] as strings e.g.

  >>> p = Poly(y*x**2, x, domain='ZZ[y]')

oscarbenjamin commented:
Looks like content and primitive are not working properly over the Gaussian domains:
```julia
In [7]: p1 = Poly(x**2/2 + 1)                                                                                                                  

In [8]: p2 = Poly(x**2/2 + I)                                                                                                                  

In [9]: p1.content()                                                                                                                           
Out[9]: 1/2

In [10]: p1.primitive()                                                                                                                        
Out[10]: (1/2, Poly(x**2 + 2, x, domain='QQ'))

In [11]: p2.content()                                                                                                                          
Out[11]: 1/16

In [12]: p2.primitive()                                                                                                                        
Out[12]: (1/16, Poly(8*x**2 + 16*I, x, domain='QQ_I'))
```

oscarbenjamin commented:
The problem with primitive is caused by a problem with gcd which is in turn caused by a problem with remainder with `QQ_I`:
```julia
In [5]: QQ_I.gcd(QQ_I(S.Half), QQ_I(0))                                                                                           
Out[5]: GaussianRational(1/4, 0)

In [7]: QQ_I(QQ(4, 7)) % QQ_I(QQ(3, 7))                                                                                           
Out[7]: GaussianRational(0, 0)
```
Currently the remainder of `GaussianRational` is always 0.

This is is inconsistent with ZZ except that ZZ is also inconsistent with itself. This is what I get with gmpy2 installed:
```julia
In [9]: QQ(4,7) % QQ(3, 7)                                                                                                        
Out[9]: mpq(1,7)
```
This is the result without gmpy2:
```julia
In [1]: QQ(4,7) % QQ(3, 7)                                                                                                        
Out[1]: 0
```
That latter result comes from PythonRational which always returns 0 like GaussianRational does.

oscarbenjamin commented:
Is it intentional that the remainder is negative in this example?
```julia
In [4]: ZZ_I(2) % ZZ_I(3)                                                                                                         
Out[4]: GaussianInteger(-1, 0)
```

jksuom commented:
> Is it intentional that the remainder is negative in this example?

It looks like the remainder is the representative having least norm. That probably follows from the division identity for Gaussian integers where the remainder has to be smaller than the divisor. There could be other possibilities for the remainder.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add factorisation over the Gaussian rationals:

  >>> factor(x**2 + x*(I - 2) - 2*I, domain='QQ_I')
  (x - 2)⋅(x + ⅈ)

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Remove unneeded assignments (flake8)

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add factorisation over the Gaussian integers:

  >>> factor(x**2 + x*(I - 2) - 2*I)
  (x - 2)⋅(x + ⅈ)

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add convert methods from various types to Gaussian domains.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add convert methods from various types to Gaussian domains.

oscarbenjamin commented:
Now in this PR `ZZ_I` and `QQ_I` are automatically detected from the coefficients without needing to specify `extension=True` e.g.:
```julia
In [26]: Poly(x**2 + I)                                                                                                                        
Out[26]: Poly(x**2 + I, x, domain='ZZ_I')

In [27]: Poly(x**2 + I/2)                                                                                                                      
Out[27]: Poly(x**2 + I/2, x, domain='QQ_I')
```
Also factorisation is implemented so that `QQ_I` is converted to `QQ<I>` in `dmp_factor_list`. This means that `factor_list` works for `QQ_I`:
```julia
In [29]: Poly(x**2/4 + 1, x, domain='QQ_I').factor_list()                                                                                      
Out[29]: (1/4, [(Poly(x - 2*I, x, domain='QQ_I'), 1), (Poly(x + 2*I, x, domain='QQ_I'), 1)])
```
In turn that means that `factor` will factorise over the Gaussian rationals:
```julia
In [34]: factor(x**2 + 1)                                                                                                                      
Out[34]: 
 2    
x  + 1

In [35]: factor(x**2 + 1, domain='QQ_I')                                                                                                       
Out[35]: (x - ⅈ)⋅(x + ⅈ)
```
This will happen automatically in factor if it happens to pick up on `I` in the expression and the expression contains rationals:
```julia
In [42]: factor(x**2 + 2*I*x/3 + S(1)/3)                                                                                                       
Out[42]: 
(x + ⅈ)⋅(3⋅x - ⅈ)
─────────────────
        3 
```
Without rationals `factor` will attempt to factor over `ZZ_I`. Factorisation over `ZZ_I` is also implemented and works by factorising over `QQ_I` and then clearing the denominators and attempting to cancel them with the coefficient so e.g.:
```julia
In [46]: factor(x**2 - I*x + 2)                                                                                                                
Out[46]: (x - 2⋅ⅈ)⋅(x + ⅈ)
```
I have found an example which fails though which is a polynomial in `ZZ_I` that can be factored in `QQ_I` but not in `ZZ_I`:
```julia
In [47]: p = 8192*x**2 + x*(22656 + 175232*I) - 921416 + 242313*I                                                                              

In [48]: p                                                                                                                                     
Out[48]: 
      2                                           
8192⋅x  + x⋅(22656 + 175232⋅ⅈ) - 921416 + 242313⋅ⅈ

In [49]: factor(p, domain='QQ_I')                                                                                                              
Out[49]: 
                       2
     ⎛    177   1369⋅ⅈ⎞ 
8192⋅⎜x + ─── + ──────⎟ 
     ⎝    128    128  ⎠ 
```
This example currently fails with factor in this PR because factor chooses to use `ZZ_I` and we get:
```julia
In [50]: factor(p)                                                                                                                             
---------------------------------------------------------------------------
CoercionFailed 
...
~/current/sympy/sympy/sympy/polys/factortools.py in dmp_factor_list(f, u, K0)
   1320         factors = new_factors
   1321         factors = [(dmp_convert(fac, u, K1, K0), i) for fac, i in factors]
-> 1322         coeff = K0.convert(coeff, K1)
   1323         return coeff, factors
   1324 
...
CoercionFailed: can't convert 1/2 of type <class 'sympy.polys.domains.pythonrational.PythonRational'> from QQ to ZZ
```
That can be fixed and I suppose that the solution in `dmp_factor_list` is that this polynomial is irreducible and primitive in `ZZ_I` so in this case we should have
```julia
>>> Poly(expr, x, domain='ZZ_I').factor_list()
(1, [(Poly(8192*x**2 + x*(22656 + 175232*I) - 921416 + 242313*I, x, domain='ZZ_I'), 1)])
```
I'm not sure in general if there were more than two factors over `QQ_I` how we should tell whether the result can be factored over `ZZ_I` e.g. if `p` above was multiplied by a linear poly in `ZZ_I` then there would be three factors and the result would be factorisable over `ZZ_I` with `p` as one of the factors e.g.:
```julia
In [59]: p2 = Poly(x - I) * p                                                                                                                  

In [60]: p2.set_domain('QQ_I').factor_list()                                                                                                   
Out[60]: (8192, [(Poly(x - I, x, domain='QQ_I'), 1), (Poly(x + 177/128 + 1369*I/128, x, domain='QQ_I'), 2)])
```
I'm not sure how to detect that and recombine the factors into those that are irreducible in `ZZ_I`.

I'm also unsure what the default behaviour of `factor` should be. Currently the PR gives:
```julia
In [62]: factor(x**2 + 1, domain='ZZ_I')                                                                                                       
Out[62]: (x - ⅈ)⋅(x + ⅈ)

In [63]: factor(I*x**2 + I)   # <--- doesn't factor
Out[63]: 
  ⎛ 2    ⎞
ⅈ⋅⎝x  + 1⎠

In [64]: factor(x**2 + 2*I*x/3 + S(1)/3)                                                                                                       
Out[64]: 
(x + ⅈ)⋅(3⋅x - ⅈ)
─────────────────
        3        

In [65]: factor(3*x**2 + 2*I*x + S(1))                                                                                                         
Out[65]: (x + ⅈ)⋅(3⋅x - ⅈ)
```
The documented behaviour of `factor` is that it factorises over the rationals. Its capability on master to factorise anything involving `I` is purely through treating `I` as a generator and factorising as a multivariate polynomial.

The effect with this PR is that factorisation over `ZZ_I` or `QQ_I` happens for any polynomial subexpression that has `I` and does/does not have rationals. This means that `Out[63]` doesn't factorise because `I` gets factored out as a common term and we end up with `Poly(x**2 + 1).factor_list()` (a poly in `ZZ` gets factored over `QQ`). This seems unsatisfactory to me and I think that maybe there should be some option for specifying whether you want to factorise over the Gaussian domains or not.

Given the factor always factorises over the rationals rather than integers by default maybe it should always choose `QQ_I` rather than `ZZ_I` by default. I'm not sure where this decision is made inside `factor` though...

So I have 4 questions:

1. Is the approach taken here for `factor_list` reasonable?
2. Is there a good way to detect exactly which of the factors should be combined into the result for `factor_list` over `ZZ_I`?
3. What should be the behaviour of `factor` when there is no domain argument?
4. Should there be an option like `gaussians=True` for factor?

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Fix factorisation over the Gaussian integers by extracting the content
from rational factors:

  >>> factor(8192*x**2 + x*(22656 + 175232*I) - 921416 + 242313*I)
  I*(x*(64 - 64*I) + 773 + 596*I)**2

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Make factor_list put pure_complex factors into the coefficient rather
than the Poly factors.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Remove unneeded import

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Use factor_terms in sign._eval_simplify because factor no longer factors
2 + 2*I.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Fix two ODE tests after changes to factor

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add tests and docstrings and clean up. Use a faster algorithm for
__pow__ in GaussianElement.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Refactor code for factorising over ZZ_I and QQ_I into
dmp/dup_factor_zz/qq_i functions in factortools.py

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Remove unneeded imports and assignments.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Remove unneeded imports and assignments.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Ensure that dup_factor_list and dmp_factor_list can still post-process
the factors when factoring over the Gaussian domains.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Only test converting from ZZ_gmpy and QQ_gmpy if gmpy is installed.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add GaussianElement.__lt__ for sorted factors.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Remove temporary workaround in apart that avoid factorising over ZZ_I
and QQ_I in favour of QQ<I>

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add tests for dup/dmp_qq/zz_i_factor functions in factortools.py.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Fix printing of PolyRing over a GaussianElement domain. The previous
code assumed that it was operating over an ordered domain and got
confused about minus signs lead to e.g.:

   >>> R, x = ring("x", ZZ_I)
   >>> print(x)
   --1*x

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Fix PythonRational.__eq__ to return NotImplemented so
PythonRational(1,1) can compare equal to Ring.one.

oscarbenjamin commented:
This is ready for review

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Add tests for high-level functions (cancel/factor) that can use the
Gaussian domains.

oscarbenjamin committed:
Merge branch 'master' into pr_gaussian_poly

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into pr_gaussian_poly

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Change test for gcd(sqrt(-2), -sqrt(-2)) so that either +-sqrt(2) is
accepted.

jksuom commented:
This looks fine. Thank you for doing this.

My only comment is about `unify`. If there is an `AlgebraicField` and a `GaussianDomain`, then the latter should probably be coerced to an `AlgebraicField` and combined with the former. But that is not essential for this PR. I think this is ready to be merged unless you would want to squash some commits.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Fix unification of ZZ_I and QQ_I with AlgebraicField, ComplexField and
RealField.

oscarbenjamin commented:
I've fixed unify with `ZZ_I/QQ_I` and algebraic fields:
```julia
In [22]: ZZ_I.unify(QQ.algebraic_field(sqrt(2)))                                                                                               
Out[22]: QQ<sqrt(2) + I>

In [23]: QQ_I.unify(QQ.algebraic_field(sqrt(2)))                                                                                               
Out[23]: QQ<sqrt(2) + I>
```
There is one outstanding issue which is what to do with the `gaussian` option in polys. Should it give `QQ_I` (or `ZZ_I`?) rather than `QQ<I>` as it currently does?
```julia
In [24]: Poly(x**2, x, gaussian=True)                                                                                                          
Out[24]: Poly(x**2, x, domain='QQ<I>')
```

oscarbenjamin commented:
I've changed it so that `gaussian=True` gives `QQ_I` rather than `QQ<I>`.

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Make the gaussian=True polys option use QQ_I rather than QQ<I>

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into pr_gaussian_poly

oscarbenjamin committed:
feat(polys): Add Gaussian domains to Poly

Update dsolve test after adding the Gaussian domains.

oscarbenjamin commented:
Thanks for reviewing @jksuom. I don't think I can do anything meaningful with squashing the commits here and this keeps going stale so I'm merging it.

oscarbenjamin commented:
I guess I should add something to the release notes...
